internal/admin/admin.go

```diff
@@ -544,35 +546,50 @@ func (adm *Admin) updateLogStream(ctx context.Context, lsid types.LogStreamID, p
 
 	clusmeta, err := adm.mrmgr.ClusterMetadataView().ClusterMetadata(ctx)
 	if err != nil {
-		return nil, err
+		return nil, status.Errorf(codes.Unavailable, "update log stream: %s", err.Error())
 	}
 
 	oldLSDesc, err := clusmeta.MustHaveLogStream(lsid)
-	if err != nil {
-		return nil, err
+	if err != nil || oldLSDesc.Status.Deleted() {
+		return nil, status.Errorf(codes.NotFound, "update log stream: no such log stream %d", lsid)
 	}
 
-	status := oldLSDesc.GetStatus()
-	if status.Running() || status.Deleted() {
-		return nil, errors.Errorf("invalid log stream status: %s", status)
+	if oldLSDesc.Status.Running() {
+		return nil, status.Errorf(codes.FailedPrecondition, "update log stream: invalid log stream status %s", oldLSDesc.Status)
 	}
-
-	replace := false
-	newLSDesc := proto.Clone(oldLSDesc).(*varlogpb.LogStreamDescriptor)
-	for i := range newLSDesc.Replicas {
-		// TODO - fix? poppedReplica can ignore path.
-		if newLSDesc.Replicas[i].GetStorageNodeID() == poppedReplica.GetStorageNodeID() {
-			newLSDesc.Replicas[i] = &pushedReplica
-			replace = true
-			break
+	popIdx, pushIdx := -1, -1
+	for idx := range oldLSDesc.Replicas {
+		snid := oldLSDesc.Replicas[idx].StorageNodeID
+		if snid == poppedReplica.StorageNodeID {
```

Comment:
We should also consider the case where popped and pushed have the same snid but different path.

Reply:
Good point!

That the popped and pushed replicas have the same StorageNodeID but different paths seem to be confused. I am not sure whether it should be considered moving the data directory to the same node or a bad request since both decisions look convincing.

In my opinion, although it might more natural to consider moving the data directory, we would better support that later, since it can have side effects on other components like storage nodes.

Reply:
Added lines to handle the same storage node.

```go
	if poppedReplica.StorageNodeID == pushedReplica.StorageNodeID {7 hours ago • Injun Song [feat(admin): use gRPC codes to clarify errors…]
		if poppedReplica.Path != pushedReplica.Path {
			return nil, status.Errorf(codes.Unimplemented, "update log stream: moving data directory")
		}
		return nil, status.Errorf(codes.InvalidArgument, "update log stream: the same replica")
	}
```

Reply:
Okay. We will modify it so that it does not make a request with a different path to the same snid.